rust/actyx/ax-aql/src/language/type_check/mod.rs

```diff
@@ -0,0 +1,159 @@
+use super::workflow::{Participant, Workflow, WorkflowStep};
+use crate::{Ident, NonEmptyVec, Query, SimpleExpr};
+use std::collections::BTreeSet;
+
+mod drill_down_type;
+
+pub(crate) struct QueryTypeCheckError<'a> {
+    pub(crate) span: pest::Span<'a>,
+    pub(crate) error_type: QueryTypeCheckErrorType,
+}
+
+pub(crate) enum QueryTypeCheckErrorType {
+    UndeclaredEvent,
+    UndeclaredParticipant,
+    UndeclaredWorkflowCall,
+    InvalidIndex(String),
+}
+struct WorkflowAnalysisResource<'a> {
+    workflow: &'a Workflow<'a>,
+    participants_set: BTreeSet<&'a Ident>,
+}
+
+impl<'a> WorkflowAnalysisResource<'a> {
+    fn from(workflow: &'a Workflow) -> Self {
+        Self {
+            workflow,
+            participants_set: workflow
+                .args
+                .iter()
+                .map(|x| {
+                    let participant: &Participant = x;
+                    match participant {
+                        Participant::Role(ident) => ident,
+                        Participant::Unique(ident) => ident,
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+pub(crate) struct QueryTypeCheck<'a> {
```

Comment:
I don't see how this structure helps the overall goal. Why not use just the `Query<'a>`

Reply:
this should have been removed. nice catch